src/mongo/db/namespace_string.h

```diff
@@ -277,6 +309,10 @@ class NamespaceString {
             : StringData(_ns.c_str() + _dotIndex + 1, _ns.size() - 1 - _dotIndex);
     }
 
+    boost::optional<TenantId> tenantId() {
```

Comment:
This allocate heap memory on every call if tenantId exists. I wonder if you need to add a `hasTenantId` that returns bool that avoids the memory allocation if callers simply need to test for tenantId.